build.zig

```diff
@@ -782,7 +782,17 @@ fn addCmakeCfgOptionsToExe(
                 mod.linkSystemLibrary("unwind", .{});
             },
             .ios, .macos, .watchos, .tvos, .visionos => {
-                mod.link_libcpp = true;
+                if (static or !std.zig.system.darwin.isSdkInstalled(b.allocator)) {
+                    mod.link_libcpp = true;
+                } else {
+                    // Avoid using `mod.linkSystemLibrary()`, which:
+                    //   - is semantically equivalent to `-lc++`
+                    //   - and enables `mod.link_libcpp`
+                    // Instead, add the full object pathname.
```

Comment:
I see, thank you. This reveals a contradiction in this implementation:
* by checking for the string "c++", `linkSystemLibrary` reveals an assumption that there is only one possible way to link libc++.
* The entire point of this patch rests on the assumption that there are multiple ways to link libc++.

Instead of working around it, the build system should be modified to recognize that there are two choices when linking libc++, because the logic that makes my code snippet not work for this use case is based on this incorrect assumption.

Then, my simpler code snippet which does not rely on something that requires access outside of a build sandbox (#14286) will work correctly.

Reply:
@andrewrk would you be okay with taking the patch as-is (since 0.14.1 is around the corner) and fixing the libc++ linking logic in the build system later?

I have on my to-do list to audit the build system's handling of all "special" libraries (libc, libc++, libunwind, ...) because I'm pretty sure there are multiple bugs lurking there besides this one. But I suspect it's going to be a bit involved to fix them, and it will likely involve breaking changes in the API, which would not be acceptable for a patch release.

Reply:
Just fyi I did start to look into doing as Andrew suggested and there are a few defs  of `link_libcpp` including `builtin`, `Compilation.Config` and `std.Build.Module` (and deprecated `std.Build`) and didn't have the bandwidth to nail that down.